compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -1587,10 +1588,23 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         unevaluated: ty::Unevaluated<'tcx>,
         span: Option<Span>,
     ) -> EvalToConstValueResult<'tcx> {
-        let mut original_values = OriginalQueryValues::default();
-        let canonical = self.canonicalize_query((param_env, unevaluated), &mut original_values);
+        let param_env = self.tcx.erase_regions(param_env);
+        let mut substs = unevaluated.substs(self.tcx);
+        substs = self.tcx.erase_regions(substs);
+        substs = self.resolve_vars_if_possible(substs);
+
+        // Postpone the evaluation of constants whose substs depend on inference
+        // variables
+        if substs.has_infer_types_or_consts() {
+            return Err(ErrorHandled::TooGeneric);
+        }
+
+        let unevaluated = ty::Unevaluated {
+            def: unevaluated.def,
+            substs_: Some(substs),
+            promoted: unevaluated.promoted,
+        };
```

Comment:
I suggest we re-order this as follows:

```suggestion
        let mut substs = unevaluated.substs(self.tcx);
        substs = self.resolve_vars_if_possible(substs);

        // Postpone the evaluation of constants whose substs depend on inference
        // variables
        if substs.has_infer_types_or_consts() {
            return Err(ErrorHandled::TooGeneric);
        }

        let param_env_erased = self.tcx.erase_regions(param_env);
        let substs_erased = self.tcx.erase_regions(substs);

        let unevaluated = ty::Unevaluated {
            def: unevaluated.def,
            substs_: Some(substs_erased),
            promoted: unevaluated.promoted,
        };

        // The return value is the evaluated value which doesn't contain any reference to inference
        // variables, thus we don't need to substitute back the original values.
        self.tcx.const_eval_resolve(param_env, unevaluated, span)
```

It is necessary to resolve variables *before* erasing because an inference variable may resolve to an `&T` of some kind.